File: Easy/palindrome.py
# Return True/False if this word is a palindrome.
# A palindrome is a word that is spelled the same backwards and forwards.


def is_palindrome(word):
    """Return True if word is palindrome, else return False.

        >>> is_palindrome("a")
        True

        >>> is_palindrome("noon")
        True

        >>> is_palindrome("racecar")
        True

        >>> is_palindrome("porcupine")
        False

        >>> is_palindrome("Racecar")
        False

    """

    i = 0
    j = -1

    while i < len(word)/2:
        if word[i] == word[j]:
            i += 1
            j -= 1
        else:
            return False
    return True
    
    # ALTERNATIVE SOLUTION
    
    # if word == word[::-1]:
    #     return True
    # else:
    #     return False

if __name__ == "__main__":
    import doctest
    doctest.testmod(verbose=True)

# Runtime O(n)

File: Easy/Ruby/palindrome.rb
# A palindrome is a word that is spelled the same backwards and forwards.

# Return true if word is palindrome, else return false.
def is_palindrome(word)
  word == word.reverse ? true : false
end

# ALTERNATIVE SOLUTION:
# This is way less ruby

# def is_palindrome(word)
#   i = 0
#   j = -1
#   while i < word.length/2
#     if word[i] == word[j]
#       i += 1
#       j -= 1
#     else
#       return false
#     end
#   end
# end

puts is_palindrome("a")
# true
puts is_palindrome("noon")
# true
puts is_palindrome("racecar")
# true
puts is_palindrome("porcupine")
# false
puts is_palindrome("Racecar")
# false